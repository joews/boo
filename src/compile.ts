import { Ast, AstNode, InstructionNode, Program } from './types'
import getOpcodeByName from './instructions'

// TODO instance state
let code: Uint8Array
let ip: number

// map labels to their decoded code address
let labels: { [label: string]: number }

// instructions can refer to labels that have not yet been defined.
// keep a map of labels to instructions that are waiting for the label's
// code address so that the address can be filled in ("backpatched") when
// we reach its declaration.
let labelForwardRefs: { [label: string]: number[] }

// AST pass to compile the given AST to machine code
export default function compile (ast: Ast): Program {
  ip = 0

  let [codeSize, stackSize, globalSize] = getCodeSize(ast)
  const codeBuffer = new ArrayBuffer(codeSize)
  code = new Uint8Array(codeBuffer)

  labels = {}
  labelForwardRefs = {}

  ast.forEach(visit)

  return {
    code,
    stackSize,
    globalSize
  }
}

// AST pass to compute the size of the compiled code
function getCodeSize (ast: Ast) {
  let codeSize = 0
  let globalSize = 0
  let currentStackSize = 0
  let maxStackSize = 0

  ast.forEach(node => {
    if (node.kind === 'instruction') {
      const { argTypes, operandTypes, resultTypes } = getOpcodeByName(node.mnemonic)

      codeSize += argTypes.length + 1
      currentStackSize -= operandTypes.length
      currentStackSize += resultTypes.length

      // TODO assert that stack length is >- 0
      maxStackSize = Math.max(maxStackSize, currentStackSize)

    } else if (node.kind === 'header') {
      globalSize = node.globals
    }
  })

  return [codeSize, maxStackSize, globalSize]
}

// AST visitor to return machine code for the given AST node
function visit (node: AstNode): void {
  switch (node.kind) {
    case 'instruction':
      return visitInstruction(node)
    case 'header':
      break
  }
}

function visitInstruction (node: InstructionNode) {
  const { opcode, argTypes } = getOpcodeByName(node.mnemonic)

  if (node.label) {
    resolveLabel(node)
  }

  const buffer = new ArrayBuffer(1 + argTypes.length)
  const view = new Uint8Array(buffer)

  view.set([opcode], 0)

  const operands = mapOperands(node)

  operands.forEach((operand: number, i: number) => {
    view.set([operand], i + 1)
  })

  code.set(view, ip)
  ip += view.length
}

// Return the byte form of the node's operands
// byte input: identity
// string/float/function input: TODO constant pool
// label input: map to label
function mapOperands (node: InstructionNode) {
  // TODO somehow switch on operand type, not instruction
  switch (node.mnemonic) {
    case 'jmp':
    case 'jne':
      return getLabelAddress(node.operands[0])
    default:
      return node.operands
  }
}

// Get the code address that `label` resolves to. If `label` has not yet been
// defined, add the current ip to the list of instructions to backpatch when
// we reach the label declaration and return a sentinel.
function getLabelAddress (label: string) {
  const address = labels[label]
  if (address) {
    return [labels[label]]
  } else {
    let refs = labelForwardRefs[label] || []
    refs.push(ip)
    labelForwardRefs[label] = refs
    return [-1]
  }
}

// Resolve the node's label to its code address (the current ip)
// If any instructions have referred to this label before now,
// backpatch their references
function resolveLabel (node: InstructionNode) {
  if (node.label) {
    labels[node.label] = ip

    const forwardRefs = labelForwardRefs[node.label]
    if (forwardRefs) {
      forwardRefs.forEach(ref => {
        code[ref + 1] = ip
      })
    }

    delete labelForwardRefs[node.label]
  }
}
