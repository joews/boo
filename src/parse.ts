import { readFileSync } from 'fs'
import { join } from 'path'
import { generate } from 'pegjs'

const parserSource = readFileSync(join(__dirname, 'boo.pegjs'), 'utf8')
const parser = generate(parserSource)

export default function parse (source: string): any {
  const ast = parser.parse(source)
  // console.log(`TRACE parse: `, JSON.stringify(ast))
  return ast
}
