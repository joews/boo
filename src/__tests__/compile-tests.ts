/* eslint-env jest */
import parse from '../parse'
import compile from '../compile'

function testCompile (source: string) {
  test(source, () => {
    const parsed = parse(source)
    console.log(JSON.stringify(parsed))
    const compiled = compile(parsed)
    expect(compiled).toMatchSnapshot()
  })
}

testCompile("iconst 3")

testCompile(`
.globals 3
iconst 3
iconst 5
iadd
print
`)
