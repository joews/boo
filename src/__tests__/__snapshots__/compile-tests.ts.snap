// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`
.globals 3
iconst 3
iconst 5
iadd
print
 1`] = `
Object {
  "code": Uint8Array [
    0,
    3,
    0,
    5,
    1,
    2,
  ],
  "globalSize": 3,
  "stackSize": 2,
}
`;

exports[`
first: iconst 1
print
second: iconst 2
third: iconst 3
iadd
 1`] = `
Object {
  "code": Uint8Array [
    0,
    1,
    2,
    0,
    2,
    0,
    3,
    1,
  ],
  "globalSize": 0,
  "stackSize": 2,
}
`;

exports[`
iconst 1
jmp skip
iconst 2
iadd
skip:
  iconst 3
iadd
 1`] = `
Object {
  "code": Uint8Array [
    0,
    1,
    5,
    7,
    0,
    2,
    1,
    0,
    3,
    1,
  ],
  "globalSize": 0,
  "stackSize": 2,
}
`;

exports[`iconst 3 1`] = `
Object {
  "code": Uint8Array [
    0,
    3,
  ],
  "globalSize": 0,
  "stackSize": 1,
}
`;
